[SourceKit] Add `CollectVariableType` request

fwcd opened the pull request:
To provide accurate support for inlay type hints in SourceKit-LSP (see apple/sourcekit-lsp#406 for details), this PR introduces a new request called `CollectVariableType`. Although similar to `CollectExpressionType`, the new request provides types for variable declarations and closure parameters specifically, together with the identifiers' ranges and a flag stating whether there already exists an explicit type annotation.

### Example

```swift
let x: Int = 3
var y = "test"
```

yields the following response:

```
<VariableTypes>
(4, 5): Int (explicit type: 1)
(19, 20): String (explicit type: 0)
</VariableTypes>
```

### Additions to the SourceKit protocol

The request looks like this:

```
{
    <key.request>:            (UID)     <source.request.variable.type>,
    <key.sourcefile>:         (string)  // Absolute path to the file.
    <key.compilerargs>:       [string*] // Array of zero or more strings for the compiler arguments,
                                        // e.g ["-sdk", "/path/to/sdk"]. If key.sourcefile is provided,
                                        // these must include the path to that file.
    [opt] <key.offset>:       (int64)   // Offset of the requested range. Defaults to zero.
    [opt] <key.length>:       (int64)   // Length of the requested range. Defaults to the entire file.
}
```

The response looks like this:

```
{
    <key.variable_type_list>: (array) [var-type-info*]   // A list of variable declarations and types
}

var-type-info ::=
{
    <key.variable_offset>:       (int64)    // Offset of a variable identifier in the source file
    <key.variable_length>:       (int64)    // Length of a variable identifier an expression in the source file
    <key.variable_type>:         (string)   // Printed type of the variable declaration
    <key.variable_type_explicit> (bool)     // Whether the declaration has an explicit type annotation
}
```

### Implementation Progress

- [x] Add `VariableTypeCollector` that traverses AST and collects variable type information
- [x] Implement the new request handler
- [x] Add tests verifying basic use-cases (local variables, fields, parameters, declarations with and without type annotations)
- [x] Add support for ranged requests (`key.offset`, `key.length`)

fwcd committed:
Add CollectVariableType request to SourceKit

- Add VariableTypeCollector
  This new SourceEntityWalker collects types from variable declarations.
- Add SwiftLangSupport::collectVariableTypes
- Implement CollectVariableType request
- Provide information about explicit types in CollectVarType
- Fix HasExplicitType in VariableTypeArray
- Fix typo
- Implement ranged CollectVariableTypes requests
- Use offset/length params for CollectVariableType in sourcekitd-test
- Address a bunch of PR suggestions
- Remove CanonicalType from VariableTypeCollector
  This turned out not to be needed (for now).
- Improve doc comment on VariableTypeCollector::getTypeOffsets
- Remove unused CanonicalTy variable
- Remove out-of-date comment
- Run clang-format on the CollectVariableType implementation
- Fix some minor style issues
- Use emplace_back while collecting variable infos
- Pass CollectVariableType range to VariableTypeCollector
- Use capitalized variable names in VariableTypeArray
  ...as recommended by the LLVM coding standards
- Use PrintOptions for type printing in VariableTypeCollector
- Return void for collectVariableType
  This seems to be cleaner stylistically.
- Avoid visiting subranges of invalid range in VariableTypeCollector
- Use std::string for type buffer in VariableTypeCollectorASTConsumer
- Use plural for PrintedType in VariableTypeArray
- Remove unused fields in VariableTypeArrayBuilder
- Add suggested doc comments to VariableTypeArray
- Remove unused VariableTypeInfo.TypeLength
- Fix typo of ostream in VariableTypeCollectorASTConsumer
- Fix typo
- Document Offset and Length semantics in collectVariableTypes

fwcd committed:
Add tests for CollectVariableType request

- Add CollectVariableType test for function/closure parameters
- Test ranged CollectVariableType request
- Add separate test for ranged CollectVariableType
- Use line:col positions in sourcekitd-test for var types
- Add test case for CollectVariableType and inout params
- Update ranged CollectVariableType test
  ...as per the PR suggestions.
- Fix style issue

ahoppen commented on tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp:
```diff
@@ -1934,6 +1949,26 @@ static void printExpressionType(sourcekitd_variant_t Info, llvm::raw_ostream &OS
+    unsigned Offset = sourcekitd_variant_dictionary_get_int64(Item, KeyVariableOffset);
+    unsigned Length = sourcekitd_variant_dictionary_get_int64(Item, KeyVariableLength);
+    bool HasExplicitType = sourcekitd_variant_dictionary_get_bool(Item, KeyVariableTypeExplicit);
+    OS << "(" << Offset << ", " << Offset + Length << "): "
```
I’m not sure what you think about this, but I think it would be easier to write tests if we were using line-column based positions here instead of offset-based ones. You can convert between the two using `resolveToLineCol`.

fwcd replied:
I feel like offset-based positions are more in line with the request itself (which uses offset + length), slightly more compact and not much harder to test. But I would be fine either way.

ahoppen replied:
I just thought that the tests read better, because e.g. [here](https://github.com/apple/swift/pull/37867/files#diff-2a7ab5b5b0f42a4779d38a6699a951ec6013673074023b669f65cac2699dc11dR32) I can’t tell at a glance what offset `230`, whereas `(19:6, 19:11)` makes it clear at first glance why variable we’re referring to.

ahoppen commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
+  /// Specified by the client whether we should canonicalize types before printing.
+  const bool CanonicalType;
+
+  /// Find an exicisting offset in the type buffer, otherwise print the type.
```
This comment actually confused me for a bit. I think the following would be better:
```cpp
/// Returns the start and end offset of this string in `OS`. If it does not exist in `OS` yet, `PrintedType` will be added to `OS`.
```

ahoppen commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
+  unsigned int BufferId;
+  std::vector<VariableTypeInfo> &Results;
+
+  /// This is where we print all types.
```
Since the entire handling of type strings is a little bit confusing at first, I think it would be good to add some background documentation here, like.

```cpp
In a SourceKit request, we transfer all type names in a single string. To describe individual types, we use offsets into that big string. `OS` builds a string that contains all null-terminated type strings. When referring to one of these types, we can use the offset at which it starts in `OS`.
```

ahoppen replied:
I thought `VariableTypeArray` used the buffer + offset to transfer the types or am I mistaken?

fwcd replied:
In `VariableTypeReader` ([`VariableTypeArray.cpp`](https://github.com/fwcd/swift/blob/43b6c0a0a27fbe9a087608d037c73ef1060777a8/tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp#L87)), the sourcekitd dictionaries are built by extracting the types at their corresponding offsets from the buffer.

ahoppen replied:
Ah, I see. In that case I *think* that we are using the output stream to avoid allocating new memory for each string. I think @nkcsgexi implemented `ExpressionTypeInfo`, maybe he’s got more background. 

In any case, I think it would be cleaner to use `StringRef`s instead of the offsets we are using at the moment.

ahoppen commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
+          Ty->reconstituteSugar(true)->print(OS);
+        }
+      }
+      auto Ty = getTypeOffsets(Buffer.str());
```
Since you already have a comment 
```cpp
// Print type type to a temporary buffer
``` 

above, I think it would be good to add another one here:
```cpp
// Transfer the type to `OS` if needed and get the offsets of this string in `OS`.
```

ahoppen commented on tools/SourceKit/include/SourceKit/Core/LangSupport.h:
```diff
@@ -138,6 +138,18 @@ struct ExpressionTypesInFile {
+struct VariableType {
+  unsigned VarOffset;
+  unsigned VarLength;
+  unsigned TypeOffset;
```
Since `TypeOffset` is not really self-explaining, could you add a doc-comment for it? Like
```cpp
/// The offset of the type’s string representation inside \c VariableTypesInFile.TypeBuffer
```

ahoppen commented on tools/SourceKit/include/SourceKit/Core/LangSupport.h:
```diff
@@ -138,6 +138,18 @@ struct ExpressionTypesInFile {
+
+struct VariableTypesInFile {
+  std::vector<VariableType> Results;
+  StringRef TypeBuffer;
```
Again, I think a doc-comment would be good here:
```cpp
/// A string containing the string representation of all types in `Results`. Entries in `Results` refer to their types by using an offset into this string.
```

ahoppen commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -206,6 +206,31 @@ namespace swift {
+  };
+
+  /// Collect type information for every variable declaration in \c SF;
+  /// All types will be printed to \c OS.
```
As you’ll see further below the use of `OS` kept confusing me. I think it would be good to clarify its interplay with `VariableTypeInfo`
```cpp
/// All types will be printed to \c OS and the type offsets of the \c VariableTypeInfos will refer into the string that backs this stream.
```

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (!applier(key, var)) return false;                \
+  } while (0)
+
+#define APPLY_ARRAY(Kind, Key)                                                 \
```
I don’t think you are using this macro.

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  // data[0] = VariableTypeArrayBuilder::funcs
+  // data[1] = custum buffer
```
Typo: custom

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,191 @@
+//===---------- VariableTypeArray.cpp -------------------------------------===//
```
I think it’s kind of unfortunate that we need to duplicate so much implementation between `ExpressionTypeArray` and `VariableTypeArray`. Do you think it would be feasible to share some implementation between the two? It might not be possible, I didn’t check…

fwcd replied:
I agree, that's a bit unfortunate. Sharing the implementation might be a bit tricky though, as many parts of this implementation make close assumptions about how `VariableType` and `ExpressionType` are structured and would require a fair bit of templating to abstract over these differences, from what I can tell.

ahoppen replied:
If that’s the case, I think that duplicating is indeed the best option…

ahoppen reviewed (commented):
Mostly looking good to me. I’ve added some questions and comments inline.

fwcd commented:
Thanks for the feedback, I've followed the implementation of `ExpressionTypeInfo` pretty closely for now, but I will make sure to address these issues.

fwcd commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
+public:
+  VariableTypeCollector(SourceFile &SF,
+                        std::vector<VariableTypeInfo> &Results,
+                        bool CanonicalType,
```
The `ExpressionTypeInfo` had one, so I added it for consistency. Perhaps in the future it might be nice to have an option for displaying inlay hints with fully qualified types?

ahoppen replied:
Honestly, I don’t think there’s much value in using fully-qualified types for inlay hints, they are just too verbose. And we can always add it later if we should discover that it’s needed.

fwcd commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -206,6 +206,31 @@ namespace swift {
+  /// Reported type of a variable declaration.
+  struct VariableTypeInfo {
+    /// The start of the variable identifier.
+    uint32_t offset;
```
Sure!

fwcd commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
+      SmallString<64> Buffer;
+      {
+        llvm::raw_svector_ostream OS(Buffer);
+        auto Ty = VD->getType()->getRValueType();
```
The `ExpressionTypeInfo` implementation uses it and I assume this is to deal with types of `inout` parameters, but I haven't looked into it too deeply yet. I'll make sure to add a test case for these. Which type printing options are you thinking of?

fwcd replied:
Ah thanks, this change seems to have been accidentally reverted with the printed-type-buffer-change

fwcd commented on tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp:
```diff
@@ -67,6 +67,7 @@ static void printNameTranslationInfo(sourcekitd_variant_t Info, llvm::raw_ostrea
 static void printRangeInfo(sourcekitd_variant_t Info, StringRef Filename,
                            llvm::raw_ostream &OS);
 static void printExpressionType(sourcekitd_variant_t Info, llvm::raw_ostream &OS);
+static void printVariableType(sourcekitd_variant_t Info, llvm::raw_ostream &OS);
```
Sure, this would deviate from how `printExpressionType` is named though, which, I assume, was directly named after the request.

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,174 @@
+  //  - Variable length in the source buffer
+  //  - The printed type
+  //  - Whether the variable has an explicit type annotation
+  CompactArrayReader<unsigned, unsigned, const char *, unsigned> EntryReader;
```
I just looked through the code how `ExpressionTypeReader` handles its strings again and I think unfortunately the output stream + offset approach is really necessary. 

This `const char *` points into `sourcekitd`’s address space, which is not readable by another process unless you are using `sourcektid` in process (which is only a debugging feature). 

The way that `ExpressionTypeReader` handles this, is by transmitting the string with all types in the buffer created by `ExpressionTypeArrayBuilder::createBuffer`. That way all strings will actually be transmitted via XPC and are thus addressable by the client process. The string with all types now has a new address (namely one in the `llvm::MemoryBuffer` returned by `createBuffer`) which gets put into the buffer’s header. That header value can then be read by [`ExpressionTypeReader`](https://github.com/apple/swift/blob/775bff003c1a41a5651e961f68a2244527b06be5/tools/SourceKit/tools/sourcekitd/lib/API/ExpressionTypeArray.cpp#L51) and all types within it can be accessed by their offset.

Sorry for the back and forth regarding this.

fwcd replied:
Thanks for digging into this, that makes sense, I'll make sure to revert those changes.

ahoppen reviewed (commented):
Sorry for the delay, I only now managed to look through the code. Unfortunately, after looking at the implementation of the new `TypeVariableArray`, I remembered why we needed the approach with the output stream + offsets in `ExpressionTypeArrayBuilder` and I think it’s also necessary for `VariableTypeArray` 😬. Sorry for the back and forth about this.

fwcd commented on test/SourceKit/VariableType/ranged.swift:
```diff
@@ -0,0 +1,8 @@
+
+// RUN: %sourcekitd-test -req=collect-var-type -pos=2:1 -length=23 %s -- %s | %FileCheck %s
+// CHECK: (2:5, 2:6): String (explicit type: 0)
+// CHECK: (3:5, 3:6): String (explicit type: 1)
```
In this case no, since we are testing ranged requests here.

ahoppen replied:
Ah, I didn't see that. Could you add the following lines
```
// CHECK-NOT: (1:5, 1:6)
```
and 
```
// CHECK-NOT: (4:5, 5:6)
```

At the moment, file check will pass even if these lines occur in the output because it will simply skip over lines that don’t match any `CHECK` line.

fwcd commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,49 @@ swift::collectExpressionType(SourceFile &SF,
+  std::vector<VariableTypeInfo> &Results;
+public:
+  VariableTypeCollector(SourceFile &SF,
+                        std::vector<VariableTypeInfo> &Results) : SM(SF.getASTContext().SourceMgr),
```
Thanks, I'll make sure to use `clang-format`. Is this a hard limit though? In some cases I find the (slightly-longer-than-80 lines) current version to be more readable than the one hard-wrapped at 80 lines. Consider:

```cpp
sourcekitd_variant_t Item =
    sourcekitd_variant_array_get_value(TypeBuffer, i);
unsigned Offset =
    sourcekitd_variant_dictionary_get_int64(Item, KeyVariableOffset);
unsigned Length =
    sourcekitd_variant_dictionary_get_int64(Item, KeyVariableLength);
auto Start = resolveToLineCol(Offset, SourceBuf);
auto End = resolveToLineCol(Offset + Length, SourceBuf);
bool HasExplicitType =
    sourcekitd_variant_dictionary_get_bool(Item, KeyVariableTypeExplicit);
auto PrintedType =
    sourcekitd_variant_dictionary_get_string(Item, KeyVariableType);
OS << "(" << Start.first << ":" << Start.second << ", " << End.first << ":"
   << End.second << "): " << PrintedType
   << " (explicit type: " << HasExplicitType << ")\n";
```

vs

```cpp
sourcekitd_variant_t Item = sourcekitd_variant_array_get_value(TypeBuffer, i);
unsigned Offset = sourcekitd_variant_dictionary_get_int64(Item, KeyVariableOffset);
unsigned Length = sourcekitd_variant_dictionary_get_int64(Item, KeyVariableLength);
auto Start = resolveToLineCol(Offset, SourceBuf);
auto End = resolveToLineCol(Offset + Length, SourceBuf);
bool HasExplicitType = sourcekitd_variant_dictionary_get_bool(Item, KeyVariableTypeExplicit);
auto PrintedType = sourcekitd_variant_dictionary_get_string(Item, KeyVariableType);
OS << "("
   << Start.first << ":" << Start.second
   << ", "
   << End.first << ":" << End.second
   << "): "
   << PrintedType
   << " (explicit type: " << HasExplicitType << ")\n";
```

ahoppen replied:
In general the files should be formatted according to `clang-format` but there’s always exceptions. Personally, I sometimes don’t like the formatting of `clang-format` sometimes and choose to overwrite it and other times people simply forget to run `clang-format`.

fwcd commented on tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp:
```diff
@@ -1934,6 +1951,34 @@ static void printExpressionType(sourcekitd_variant_t Info, llvm::raw_ostream &OS
+       << ", "
+       << End.first << ":" << End.second
+       << "): "
+       << sourcekitd_variant_dictionary_get_string(Item, KeyVariableType)
```
No reason, thanks for pointing this out, I'll use a variable!

fwcd commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -2368,3 +2368,59 @@ void SwiftLangSupport::collectExpressionTypes(StringRef FileName,
+      auto *SF = AstUnit->getCompilerInstance().getPrimarySourceFile();
+      std::vector<VariableTypeInfo> Scratch;
+      VariableTypesInFile Result;
+      for (auto Item : collectVariableType(*SF, Scratch)) {
```
Thanks for the suggestion, we still need the loop though to convert from IDETypeChecking's `VariableTypeInfo` to SourceKit's `VariableType` (even though those structures have the same fields).

ahoppen commented on test/SourceKit/VariableType/ranged.swift:
```diff
@@ -0,0 +1,8 @@
+let z: String = "def"
+var w = 4
+
+// RUN: %sourcekitd-test -req=collect-var-type -pos=2:1 -length=23 %s -- %s | %FileCheck %s
```
I *think* you can also use `-end-pos` here instead of `-length`. That would make the test case more resistent if someone chose to rename one of the variables/types/values

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,177 @@
+  CompactArrayBuilder<unsigned, unsigned, unsigned, unsigned> Builder;
+  CompactArrayBuilder<StringRef> StrBuilder;
+
+  Implementation(StringRef PrintedType) { StrBuilder.addEntry(PrintedType); }
```
Shouldn’t this be `PrintedTypes` (plural) instead of just `PrintedType`?

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,177 @@
+} // end of anonymous namespace
+
+struct VariableTypeArrayBuilder::Implementation {
+  StringRef PrintedType;
+  SmallVector<char, 256> Buffer;
```
I don’t think `PrintedType` or `Buffer` are ever used.

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,177 @@
+  // Four unsigned integers for:
+  //  - Variable offset in the source buffer
+  //  - Variable length in the source buffer
+  //  - Offset of printed Variable type inside printedType
```
Nitpick: The `printedType` variable is capitalized: `printedType` -> `PrintedType`

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,177 @@
+
+namespace {
+class VariableTypeReader {
+  const char *PrintedType;
```
Nitpick: Shouldn’t this be `PrintedTypes` (plural)?

Suggested doc comment:
```cpp
/// A string that contains the types of the variables that are reported by this \c VariableTypeReader. Each type is null-terminated and \c EntryReader references the types using offsets into this string.
```

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,177 @@
+  StringRef PrintedType;
+  SmallVector<char, 256> Buffer;
+  CompactArrayBuilder<unsigned, unsigned, unsigned, unsigned> Builder;
+  CompactArrayBuilder<StringRef> StrBuilder;
```
Suggested doc comment.

```cpp
/// A builder that builds the \c PrintedTypes string used by \c VariableTypeReader. See \c VariableTypeReader::PrintedTypes for more info.
```

ahoppen commented on tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp:
```diff
@@ -0,0 +1,177 @@
+struct VariableTypeArrayBuilder::Implementation {
+  StringRef PrintedType;
+  SmallVector<char, 256> Buffer;
+  CompactArrayBuilder<unsigned, unsigned, unsigned, unsigned> Builder;
```
Suggested doc comment

```cpp
/// A builder that builds the that builds values read by \c EntryReader in \c VariableTypeReader. See \c VariableTypeReader::EntryReader for more info.
```

ahoppen reviewed (commented):
And I’ve got some more comments inline.

fwcd commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -206,6 +206,36 @@ namespace swift {
+    uint32_t TypeOffset;
+
+    /// The length of the printed type in a separate string buffer.
+    uint32_t TypeLength;
```
If we avoid the conversion and use `VariableTypeInfo` in `LangSupport.h`, wouldn't that introduce a dependency on `IDETypeChecking`?

ahoppen replied:
True. We could move the `VariableTypeInfo` type further into a module that’s imported by both `Sema` and `SourceKit` (e.g. `AST`) but that seems like an overkill to me. Still, I think it would be nice if both types had the same layout.

fwcd commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -206,6 +206,36 @@ namespace swift {
+  /// stream.
+  ArrayRef<VariableTypeInfo> collectVariableType(
+      SourceFile &SF, Optional<unsigned> Offset, Optional<unsigned> Length,
+      std::vector<VariableTypeInfo> &Scratch, llvm::raw_ostream &OS);
```
Since we are passing through the vector anyway, I think making it return void and turning `Scratch` into a 'pure' output parameter would be good.

fwcd commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -206,6 +206,36 @@ namespace swift {
+  /// \c VariableTypeInfos will index into the string that backs this
+  /// stream.
+  ArrayRef<VariableTypeInfo> collectVariableType(
+      SourceFile &SF, Optional<unsigned> Offset, Optional<unsigned> Length,
```
Just curious, is there a reason to use `SourceRange` over `CharSourceRange` in this case, which is already used in `walkToDeclPre`?

ahoppen replied:
I chose `SourceRange` because all AST nodes already store their `SourceRange` but not their `CharSourceRange`. If you want to convert `SourceRange` to `CharSourceRange`, you need to reach out to the lexer again to determine the length of the end token. 

Thus, I think it’s better to just do the conversion once for the requested range and avoid converting the range of every AST node.

fwcd commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,96 @@ swift::collectExpressionType(SourceFile &SF,
+                        llvm::raw_ostream &OS)
+      : SM(SF.getASTContext().SourceMgr), BufferId(*SF.getBufferID()),
+        TotalOffset(Offset), TotalLength(Length), Results(Results), OS(OS) {}
+
```
Wouldn't most statements, expressions and patterns occur in declarations anyway?

ahoppen replied:
Not really, simple function calls and variable assignments don’t occur in declarations. Same goes for control flow constructs like `if` or `while`.

For example if I have
```
func foo() {
  let x = 1
  x += 1
  // repeat `x += 1` 500 times
}
```

and I only request the variable types of line 2, then there’s no need to visit the 500 `x += 1` statements.

fwcd replied:
Yes, fair point.

fwcd committed:
Migrate VariableTypeCollector to SourceRange

- Add SourceRange::contains and SourceRange::overlaps
- Use SourceRange in VariableTypeCollector

fwcd commented:
I have updated the implementation to use `SourceRange`s as suggested and will now rebase the branch on main to resolve the merge conflict and squash related commits.

ahoppen commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,112 @@ swift::collectExpressionType(SourceFile &SF,
+        Options.SynthesizeSugarOnTypes = true;
+        VD->getType()->print(OS, Options);
+      }
+      // Transfer the type to `OS` if needed and get the offsets of this string
```
`offsets` -> `offset` if `getTypeOffsets` becomes `getTypeOffset`.

ahoppen commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,112 @@ swift::collectExpressionType(SourceFile &SF,
+    : Offset(Offset), Length(Length), HasExplicitType(HasExplicitType),
+      TypeOffset(TypeOffset) {}
+
+void swift::collectVariableType(SourceFile &SF, SourceRange Range,
```
Also here, I think `SF` could be a `const` reference.

ahoppen reviewed (commented):
I’ve got some minor comments regarding variable names and doc comments again. Other than that, it’s looking great. Can’t wait to try this out with the editor.

fwcd committed:
Ignore error types in VariableTypeCollector

Also add a test to verify that error types get ignored (as intended).

fwcd commented on lib/IDE/IDETypeChecking.cpp:
```diff
@@ -731,6 +731,112 @@ swift::collectExpressionType(SourceFile &SF,
+  }
+
+public:
+  VariableTypeCollector(SourceFile &SF, SourceRange Range,
```
`SourceEntityWalker::walk` unfortunately requires a non-const reference

ahoppen replied:
Fair enough. In that case I’m fine with keeping it as-is.

fwcd committed:
Address stylistic suggestions in VariableTypeCollector

Rename VariableTypeCollector::getTypeOffsets to ::getTypeOffset and only
return the start offset, since the end offset is no longer needed (the
string is null-terminated).

Also improve variable naming slightly.

ahoppen reviewed (approved):
Nice! Let’s get this 🚢ed.

ahoppen commented:
@swift-ci Please smoke test
